src/core/eval.ml

```diff
@@ -310,51 +336,52 @@ and tree_walk : dtree -> stack -> (term * stack) option = fun tree stk ->
   walk tree stk 0 VarMap.empty IntMap.empty
 
 (** [snf t] computes the strong normal form of the term [t]. *)
-and snf : term -> term = fun t ->
-  let h = whnf t in
+and snf : ctxt -> term -> term = fun ctx t ->
+  let h = whnf ctx t in
   match h with
   | Vari(_)     -> h
   | Type        -> h
   | Kind        -> h
   | Symb(_)     -> h
+  | LLet(t,_,b) -> snf ctx (Bindlib.subst b t)
   | Prod(a,b)   ->
       let (x,b) = Bindlib.unbind b in
-      let b = snf b in
+      let b = snf ctx b in
       let b = Bindlib.unbox (Bindlib.bind_var x (lift b)) in
-      Prod(snf a, b)
+      Prod(snf ctx a, b)
   | Abst(a,b)   ->
       let (x,b) = Bindlib.unbind b in
-      let b = snf b in
+      let b = snf ctx b in
       let b = Bindlib.unbox (Bindlib.bind_var x (lift b)) in
-      Abst(snf a, b)
-  | Appl(t,u)   -> Appl(snf t, snf u)
-  | Meta(m,ts)  -> Meta(m, Array.map snf ts)
+      Abst(snf ctx a, b)
+  | Appl(t,u)   -> Appl(snf ctx t, snf ctx u)
+  | Meta(m,ts)  -> Meta(m, Array.map (snf ctx) ts)
   | Patt(_,_,_) -> assert false
   | TEnv(_,_)   -> assert false
   | Wild        -> assert false
   | TRef(_)     -> assert false
 
 (** [whnf t] computes a weak head-normal form of [t]. *)
-let whnf : term -> term = fun t ->
+let whnf : ctxt -> term -> term = fun ctx t ->
   Pervasives.(steps := 0);
   let t = unfold t in
-  let u = whnf t in
+  let u = whnf ctx t in
   if Pervasives.(!steps = 0) then t else u
 
 (** [simplify t] reduces simple redexes of [t]. *)
 let rec simplify : term -> term = fun t ->
```

Comment:
thanks to add the ctxt argument here too

in proof.ml where this is used, you can turn the environment goal_hyps into a context

in env.ml, it is used only on metavariable types that are closed. so it is justified to use [] there